File: code/ios/Classes/YcProductPlugin.swift
import Flutter
import UIKit
import YCProductSDK
import JL_BLEKit
import JLDialUnit
import RTKLEFoundation
import RTKOTASDK
import iOSDFULibrary



public class YcProductPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    
    /// ecg算法工具类
    public var ecgAlgorithmManager = YCECGManager()
    
    /// 当前ECG数据的前一个
    public var previousData: Float = 0
    
    /// 当前ECG数据的上一个的上一个
    public var previousPreviousData: Float = 0
    
    /// ECG数量
    public var ecgDataCount: Int = 0
    
    public override init() {
        super.init()
        
      // 只监听一次
      setupDeviceObserver()
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
    
//    static var eventChannel: FlutterEventChannel?
    var eventSink: FlutterEventSink?
    
    public static var methodChannel: FlutterMethodChannel? = nil
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        
        let instance = YcProductPlugin()
        
        // methodChannel
        let methodChannel = FlutterMethodChannel(name: "ycaviation.com/yc_product_plugin_method_channel", binaryMessenger: registrar.messenger())
         
        registrar.addMethodCallDelegate(
            instance,
            channel: methodChannel
        )
        
        // eventChannel
        let eventChannel = FlutterEventChannel(
            name: "ycaviation.com/yc_product_plugin_event_channel",
            binaryMessenger: registrar.messenger()
        )
        
        eventChannel.setStreamHandler(instance)
        
        self.methodChannel = methodChannel
    }
    
    // 方法调用
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        
        matchMethodCall(call, result: result)
    }
    
    // MARK: - 监听
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        
        eventSink = events
        
//        setupDeviceObserver()
        
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        
        eventSink = nil
        return nil
    }
}


// MARK: - eventChannel 通道
extension YcProductPlugin {
    
    /// 设置监听
    private func setupDeviceObserver() {
        
        NotificationCenter.default.removeObserver(self)
        
        // 监听蓝牙状态
        setupDeviceStateObserver()
        
        // 监听设备操作状态
        setupDeviceControlObserver()
        
        // 监听实时数据
        setupDeviceRealDataObserver()
        
        // 监听杰理语音识别数据
        setupJLAudioDataObserver()
    }
    
    /// 转换状态码
    /// - Parameter state: <#state description#>
    /// - Returns: <#description#>
    public func convertFlutterState(_ state: YCProductState) -> Int {
         
        if state == .noRecord ||
            state == .succeed {
            
            return PluginState.succeed
            
        } else if state == .unavailable {
            
            return PluginState.unavailable
            
        } else {
            
            return PluginState.failed
        }
        
    }
}

// MARK: - methodChannel 通信
extension YcProductPlugin {
    
    /// 处理不同的方法
    /// - Parameters:
    ///   - call: <#call description#>
    ///   - result: <#result description#>
    private func matchMethodCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        
        switch call.method {
            
            // MARK: - 初始化
        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)
            
        case "initPlugin":
            initSDK(call.arguments, result: result)
            
        case "setReconnectEnabled":
            setReconnectEnabled(call.arguments, result: result)
            
        case "clearQueue":
            clearQueue(result: result)
            
        case "getBluetoothState":
            getBluetoothState(result: result)
            
        case "scanDevice":
            scanDevice(call.arguments, result: result)
            
        case "stopScanDevice":
            stopScanDevice(result: result)
            
        case "connectDevice":
            connectDevice(call.arguments, result: result)
            
        case "disconnectDevice":
            disConnectDevice(call.arguments, result: result)
            
        case "getDeviceFeature":
            getDeviceFeature(call.arguments, result: result)
            
            // MARK: - 查询与删除健康历数据
            
        case "queryDeviceHealthData":
            queryDeviceHealthData(call.arguments, result: result)
            
        case "deleteDeviceHealthData":
            deleteDeviceHealthData(call.arguments, result: result)
            
            // MARK: - 查询部分
            
        case "queryDeviceBasicInfo":
            queryDeviceBasicInfo(call.arguments, result: result)
            
        case "queryDeviceMacAddress":
            queryDeviceMacAddress(call.arguments, result: result)
            
        case "queryDeviceModel":
            queryDeviceModel(call.arguments, result: result)
            
        case "queryDeviceMCU":
            queryDeviceMCU(call.arguments, result: result)
            
            // MARK: - 设置部分
        case "setDeviceSyncPhoneTime":
            setDeviceSyncPhoneTime(call.arguments, result: result)
            
        case "setDeviceStepGoal":
            setDeviceStepGoal(call.arguments, result: result)
            
        case "setDeviceSleepGoal":
            setDeviceSleepGoal(call.arguments, result: result)
            
        case "setDeviceUserInfo":
            setDeviceUserInfo(call.arguments, result: result)
            
        case "setDeviceSkinColor":
            setDeviceSkinColor(call.arguments, result: result)
            
        case "setDeviceUnit":
            setDeviceUnit(call.arguments, result: result)
            
        case "setDeviceAntiLost":
            setDeviceAntiLost(call.arguments, result: result)
            
        case "setDeviceNotDisturb":
            setDeviceNotDisturb(call.arguments, result: result)
            
        case "setDeviceLanguage":
            setDeviceLanguage(call.arguments, result: result)
            
        case "setDeviceSedentary":
            setDeviceSedentary(call.arguments, result: result)
            
        case "setDeviceWearingPosition":
            setDeviceWearingPosition(call.arguments, result: result)
            
        case "setPhoneSystemInfo":
            setPhoneSystemInfo(call.arguments, result: result)
            
        case "restoreFactorySettings":
            restoreFactorySettings(call.arguments, result: result)

        case "settingGetAllAlarm":
            settingGetAllAlarm(call.arguments,result:result)

        case "settingAddAlarm":
            settingAddAlarm(call.arguments,result:result)

        case "settingModfiyAlarm":
            settingModfiyAlarm(call.arguments,result:result)        

            
        case "setDeviceWristBrightScreen":
            setDeviceWristBrightScreen(call.arguments, result: result)
            
        case "setDeviceDisplayBrightness":
            setDeviceDisplayBrightness(call.arguments, result: result)
            
        case "setDeviceHealthMonitoringMode":
            setDeviceHealthMonitoringMode(call.arguments, result: result)
            
        case "setDeviceTemperatureMonitoringMode":
            setDeviceTemperatureMonitoringMode(call.arguments, result: result)
            
        case "setDeviceHeartRateAlarm":
            setDeviceHeartRateAlarm(call.arguments, result: result)
            
        case "setDeviceBloodPressureAlarm":
            setDeviceBloodPressureAlarm(call.arguments, result: result)
            
        case "setDeviceBloodOxygenAlarm":
            setDeviceBloodOxygenAlarm(call.arguments, result: result)
            
        case "setDeviceRespirationRateAlarm":
            setDeviceRespirationRateAlarm(call.arguments, result: result)
            
        case "setDeviceTemperatureAlarm":
            setDeviceTemperatureAlarm(call.arguments, result: result)
            
        case "queryDeviceTheme":
            queryDeviceTheme(call.arguments, result: result)
            
        case "setDeviceTheme":
            setDeviceTheme(call.arguments, result: result)
            
        case "setDeviceSleepReminder":
            setDeviceSleepReminder(call.arguments, result: result)
            
        case "setDeviceInfoPush":
            setDeviceInfoPush(call.arguments, result: result)
            
        case "setDevicePeriodicReminderTask":
            setDevicePeriodicReminderTask(
                call.arguments,
                result: result
            )
            
            // MARK: - App Control
            
        case "sendPhoneUUIDToDevice":
            sendPhoneUUIDToDevice(call.arguments, result: result)
            
        case "sendDeviceMenstrualCycle":
            sendDeviceMenstrualCycle(
                call.arguments,
                result: result
            )
            
        case "findDevice":
            findDevice(call.arguments, result: result)
            
        case "deviceSystemOperator":
            deviceSystemOperator(call.arguments, result: result)
            
        case "bloodPressureCalibration":
            bloodPressureCalibration(call.arguments, result: result)
            
        case "temperatureCalibration":
            temperatureCalibration(call.arguments, result: result)
            
        case "bloodGlucoseCalibration":
            bloodGlucoseCalibration(call.arguments, result: result)
            
        case "sendTodayWeather":
            sendTodayWeather(call.arguments, result: result)

        case "waveDataUpload":
            waveDataUpload(call.arguments, result: result)

            
        case "sendTomorrowWeather":
            sendTomorrowWeather(call.arguments, result: result)
            
        case "uricAcidCalibration":
            uricAcidCalibration(call.arguments, result: result)
            
        case "bloodFatCalibration":
            bloodFatCalibration(call.arguments, result: result)
            
        case "sendBusinessCard":
            sendBusinessCard(call.arguments, result: result)
            
        case "queryBusinessCard":
            queryBusinessCard(call.arguments, result: result)
            
        case "appControlTakePhoto":
            appControlTakePhoto(call.arguments, result: result)
            
        case "appControlSport":
            appControlSport(call.arguments, result: result)
            
        case "appControlMeasureHealthData":
            appControlMeasureHealthData(call.arguments, result: result)
            
        case "realTimeDataUpload":
            realTimeDataUpload(call.arguments, result: result)
            
        case "startECGMeasurement":
            startECGMeasurement(call.arguments, result: result)
            
        case "stopECGMeasurement":
            stopECGMeasurement(call.arguments, result: result)
            
        case "getECGResult":
            getECGResult(call.arguments, result: result)
            
        case "queryCollectDataBasicInfo":
            queryCollectDataBasicInfo(call.arguments, result: result)
            
        case "queryCollectDataInfo":
            queryCollectDataInfo(call.arguments, result: result)
            
        case "deleteCollectData":
            deleteCollectData(call.arguments, result: result)
            
            
            // MARK: - Ota
        case "deviceUpgrade":
            deviceUpgrade(call.arguments, result: result)
            
            // MARK: - 表盘操作
        case "queryWatchFaceInfo":
            queryWatchFaceInfo(call.arguments, result: result)
            
        case "changeWatchFace":
            changeWatchFace(call.arguments, result: result)
            
        case "deleteWatchFace":
            deleteWatchFace(call.arguments, result: result)
            
        case "installWatchFace":
            installWatchFace(call.arguments, result: result)
            
        case "queryDeviceCustomWatchFaceInfo":
            queryDeviceCustomWatchFaceInfo(call.arguments, result: result)
            
        case "installCustomWatchFace":
            installCustomWatchFace(call.arguments, result: result)
            
        case "changeJieLiWatchFace":
            changeJieLiWatchFace(call.arguments, result: result)
            
        case "deleteJieLiWatchFace":
            deleteJieLiWatchFace(call.arguments, result: result)
            
        case "installJieLiWatchFace":
            installJieLiWatchFace(call.arguments, result: result)
            
            
        case "queryDeviceDisplayParametersInfo":
            queryDeviceDisplayParametersInfo(
                call.arguments,
                result: result
            )
            
        case "installJieLiCustomWatchFace":
            installJieLiCustomWatchFace(
                call.arguments,
                result: result
            )

         case "writeImageToJLDevice":
            writeImageToJLDevice(
                call.arguments,
                result: result
            )
            
        case "queryJieLiDeviceContacts":
            queryJieLiDeviceContacts(
                call.arguments,
                result: result
            )
            
        case "updateJieLiDeviceContacts":
            updateJieLiDeviceContacts(
                call.arguments,
                result: result
            )
            
        case "updateDeviceContacts":
            updateDeviceContacts(
                call.arguments,
                result: result
            )

        case "getLogFilePath":
            getLogFilePath(call.arguments,
                result: result)    

        case "getJLDeviceLogFilePath":
            getJLDeviceLogFilePath(call.arguments,
                result: result)
                
        case "getDeviceLogFilePath":
            getDeviceLogFilePath(call.arguments,
                result: result)

         case "clearSDKLog":
                    clearSDKLog(call.arguments,
                        result: result)
            
        case "sendDeviceQuiteSleep":
            sendDeviceQuiteSleep(result: result)
            
            

        // MARK: - 语音识别
        case "startSpeechRecognition":
            startSpeechRecognition(result: result)
            
        case "stopSpeechRecognition":
            stopSpeechRecognition(result: result)
            
            
        default:
            result(FlutterMethodNotImplemented)
        }
         
    }
}
 

// MARK: - 工具代码
extension YcProductPlugin {
    
    /// 圆角图片
    public static func circleImage(
        _ srcImage: UIImage,
        borderColor: UIColor,
        borderWidth: CGFloat,
        corner: CGFloat,
        finalSize: CGSize
    ) -> UIImage? {
        
        let scale = UIScreen.main.scale
        UIGraphicsBeginImageContextWithOptions(finalSize, false, scale)
        
        let path = UIBezierPath(roundedRect: CGRect(origin: .zero, size: finalSize), cornerRadius: corner)
        
        path.lineWidth = borderWidth
        borderColor.set()
        path.addClip()
        path.fill()
        
        let clicPath =
            UIBezierPath(
                roundedRect:
                    CGRect(x: borderWidth, y: borderWidth, width: finalSize.width - 2 * borderWidth, height: finalSize.height - 2 * borderWidth),
                cornerRadius: corner
            )
        
        clicPath.addClip()
        
        srcImage.draw(in: CGRect(x: borderWidth, y: borderWidth, width: finalSize.width - 2 * borderWidth, height: finalSize.height - 2 * borderWidth))
        
        let newImage = UIGraphicsGetImageFromCurrentImageContext()
        
        UIGraphicsEndImageContext()
        
        return newImage
    }
}

File: code/android/src/main/java/com/example/yc_product_plugin/YcProductPlugin.java
package com.example.yc_product_plugin;

import android.content.Context;
import android.app.Activity;
import android.content.Intent;
import android.os.Handler;
import android.util.Log;

import androidx.annotation.NonNull;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;


/** YcProductPlugin */
public class YcProductPlugin implements FlutterPlugin, ActivityAware, MethodCallHandler {

  private MethodChannel methodChannel;
  private EventChannel eventChannel;

  // 定义一个上下文
  private Context context;
  private Handler handler;
  private EventChannel.EventSink eventSink;

  private ActivityPluginBinding activityPluginBinding;


  @Override
  public void onAttachedToActivity(@NonNull ActivityPluginBinding binding) {
    activityPluginBinding = binding;
  }

  @Override
  public void onDetachedFromActivityForConfigChanges() {

  }

  @Override
  public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding) {

  }

  @Override
  public void onDetachedFromActivity() {

  }

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {

    // 初始化
    context = flutterPluginBinding.getApplicationContext();
    handler = new Handler();


    // methodChannel
    methodChannel =
            new MethodChannel(
                    flutterPluginBinding.getBinaryMessenger(),
                    "ycaviation.com/yc_product_plugin_method_channel"
            );
    methodChannel.setMethodCallHandler(this);

    // eventChannel
    eventChannel =
            new EventChannel(
                    flutterPluginBinding.getBinaryMessenger(),
                    "ycaviation.com/yc_product_plugin_event_channel"
            );


    eventChannel.setStreamHandler(new EventChannel.StreamHandler() {

      @Override
      public void onListen(Object arguments, EventChannel.EventSink events) {

        eventSink = events;
        setupObserver(handler, eventSink);

      }

      @Override
      public void onCancel(Object arguments) {

      }
    });
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    matchMethodCall(call, result);
  }


  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    // 销毁通道
    methodChannel.setMethodCallHandler(null);
    eventChannel.setStreamHandler(null);
  }


  /**
   * 适配MethodChannel
   *
   * @param call
   * @param result
   */
  private void matchMethodCall(@NonNull MethodCall call, @NonNull Result result) {

    String method = call.method;
    Object arguments = call.arguments;


    switch (method) {

      // MARK: - SDK 初始化
      case "getPlatformVersion":
        result.success("Android " + android.os.Build.VERSION.RELEASE);
        break;

      case "initPlugin":
        YcProductPluginInit.initPlugin(context, arguments, result);
        break;

      case "setReconnectEnabled":
        YcProductPluginInit.setReconnectEnabled(arguments, result);
        break;

      case "clearQueue":
        YcProductPluginInit.clearQueue(result);
        break;

      case "resetBond":
        YcProductPluginInit.resetBond(arguments,result);
        break;

      case "exitScanDevice":
        YcProductPluginInit.exitScanDevice(arguments,result);
        break;

      // MARK: - 蓝牙相关

      case "getBluetoothState":
        YcProductPluginDevice.getBluetoothState(result);
        break;

      case "scanDevice":
        YcProductPluginDevice.scanDevice(context,arguments, result);
        break;

      case "stopScanDevice":
        YcProductPluginDevice.stopScanDevice(result);
        break;

      case "connectDevice":
        YcProductPluginDevice.connectDevice(arguments, result);
        break;

      case "disconnectDevice":
        YcProductPluginDevice.disConnectDevice(arguments, result);
        break;

      case "getDeviceFeature":
        YcProductPluginQuery.getDeviceFeature(arguments, result);
        break;


      // MARK: - 同步数据
      case "queryDeviceHealthData":
        YcProductPluginHealthData.queryDeviceHealthData(arguments, result);
      break;

      case "deleteDeviceHealthData":
        YcProductPluginHealthData.deleteDeviceHealthData(arguments, result);
      break;

      // MARK: - 查询
      case "queryDeviceBasicInfo":
        YcProductPluginQuery.queryDeviceBasicInfo(arguments, result);
        break;

      case "queryDeviceMacAddress":
        YcProductPluginQuery.queryDeviceMacAddress(arguments, result);
        break;

      case "queryDeviceModel":
        YcProductPluginQuery.queryDeviceModel(arguments, result);
        break;

      case "queryDeviceMCU":
        YcProductPluginQuery.queryDeviceMCU(arguments, result);
        break;

        // MARK: - 设置部分
      case "setDeviceSyncPhoneTime":
        YcProductPluginSetting.setDeviceSyncPhoneTime(arguments, result);
        break;

      case "setDeviceStepGoal":
        YcProductPluginSetting.setDeviceStepGoal(arguments, result);
        break;

      case "setDeviceSleepGoal":
        YcProductPluginSetting.setDeviceSleepGoal(arguments, result);
        break;

      case "setDeviceUserInfo":
        YcProductPluginSetting.setDeviceUserInfo(arguments, result);
        break;

      case "setDeviceSkinColor":
        YcProductPluginSetting.setDeviceSkinColor(arguments, result);
        break;

      case "setDeviceUnit":
        YcProductPluginSetting.setDeviceUnit(arguments, result);
        break;

      case "setDeviceAntiLost":
        YcProductPluginSetting.setDeviceAntiLost(arguments, result);
        break;

      case "setDeviceNotDisturb":
        YcProductPluginSetting.setDeviceNotDisturb(arguments, result);
        break;

      case "setDeviceLanguage":
        YcProductPluginSetting.setDeviceLanguage(arguments, result);
        break;

      case "setDeviceSedentary":
        YcProductPluginSetting.setDeviceSedentary(arguments, result);
        break;

      case "setDeviceWearingPosition":
        YcProductPluginSetting.setDeviceWearingPosition(arguments, result);

      case "setPhoneSystemInfo":
        YcProductPluginSetting.setPhoneSystemInfo(arguments, result);
        break;

      case "restoreFactorySettings":
        YcProductPluginSetting.restoreFactorySettings(arguments, result);
        break;

      case "setDeviceWristBrightScreen":
        YcProductPluginSetting.setDeviceWristBrightScreen(arguments, result);
        break;

      case "setDeviceDisplayBrightness":
        YcProductPluginSetting.setDeviceDisplayBrightness(arguments, result);
        break;

      case "setDeviceHealthMonitoringMode":
        YcProductPluginSetting.setDeviceHealthMonitoringMode(arguments, result);
        break;

      case "setDeviceTemperatureMonitoringMode":
        YcProductPluginSetting.setDeviceTemperatureMonitoringMode(arguments, result);
        break;

      case "setDeviceHeartRateAlarm":
        YcProductPluginSetting.setDeviceHeartRateAlarm(arguments, result);
        break;

      case "setDeviceBloodPressureAlarm":
        YcProductPluginSetting.setDeviceBloodPressureAlarm(arguments, result);
        break;

      case "setDeviceBloodOxygenAlarm":
        YcProductPluginSetting.setDeviceBloodOxygenAlarm(arguments, result);
        break;

      case "setDeviceRespirationRateAlarm":
        YcProductPluginSetting.setDeviceRespirationRateAlarm(arguments, result);
        break;

      case "setDeviceTemperatureAlarm":
        YcProductPluginSetting.setDeviceTemperatureAlarm(arguments, result);
        break;

      case "queryDeviceTheme":
        YcProductPluginSetting.queryDeviceTheme(arguments, result);
        break;

      case "setDeviceTheme":
        YcProductPluginSetting.setDeviceTheme(arguments, result);
        break;

      case "setDeviceSleepReminder":
        YcProductPluginSetting.setDeviceSleepReminder(arguments, result);
        break;

      case "setDevicePeriodicReminderTask":
        YcProductPluginSetting.setDevicePeriodicReminderTask(arguments, result);
        break;

      case "setDeviceInfoPush":
        YcProductPluginSetting.setDeviceInfoPush(arguments, result);
        break;


      // MARK: - App Control部分

      case "sendPhoneUUIDToDevice":
        YcProductPluginAppControl.sendPhoneUUIDToDevice(arguments, result);
        break;

      case "sendDeviceMenstrualCycle":
        YcProductPluginAppControl.sendDeviceMenstrualCycle(arguments, result);
        break;

      case "findDevice":
        YcProductPluginAppControl.findDevice(arguments, result);
        break;

      case "deviceSystemOperator":
        YcProductPluginAppControl.deviceSystemOperator(arguments, result);
        break;

      case "bloodPressureCalibration":
        YcProductPluginAppControl.bloodPressureCalibration(arguments, result);
        break;

      case "temperatureCalibration":
        YcProductPluginAppControl.temperatureCalibration(arguments, result);
        break;

      case "bloodGlucoseCalibration":
        YcProductPluginAppControl.bloodGlucoseCalibration(arguments, result);
        break;

      case "sendTodayWeather":
        YcProductPluginAppControl.sendTodayWeather(arguments, result);
        break;
      case"waveDataUpload":
        YcProductPluginAppControl.waveDataUpload(arguments, result);
      break;
      case "sendTomorrowWeather":
        YcProductPluginAppControl.sendTomorrowWeather(arguments, result);
        break;

      case "uricAcidCalibration":
        YcProductPluginAppControl.uricAcidCalibration(arguments, result);
        break;

      case "bloodFatCalibration":
        YcProductPluginAppControl.bloodFatCalibration(arguments, result);
        break;

      case "appPushNotifications":
        YcProductPluginAppControl.appPushNotifications(arguments, result);
        break;

      case "sendBusinessCard":
        YcProductPluginAppControl.sendBusinessCard(arguments, result);
        break;

      case "queryBusinessCard":
        YcProductPluginAppControl.queryBusinessCard(arguments, result);
        break;

      case "appControlTakePhoto":
        YcProductPluginDeviceControl.appControlTakePhoto(arguments, result);
        break;

      case "appControlSport":
        YcProductPluginDeviceControl.appControlSport(arguments, result);
        break;

      case "appControlMeasureHealthData":
        YcProductPluginDeviceControl.appControlMeasureHealthData(arguments, result);
        break;

      case "realTimeDataUpload":
        YcProductPluginDeviceControl.realTimeDataUpload(arguments, result);
        break;

      case "startECGMeasurement":
        YcProductPluginECG.startECGMeasurement(arguments, result, handler, eventSink);
        break;

      case "stopECGMeasurement":
        YcProductPluginECG.stopECGMeasurement(arguments, result);
        break;

      case "getECGResult":
        YcProductPluginECG.getECGResult(arguments, result);
        break;

      case "queryCollectDataBasicInfo":
        YcProductPluginCollectData.queryCollectDataBasicInfo(arguments, result);
        break;

      case "queryCollectDataInfo":
        YcProductPluginCollectData.queryCollectDataInfo(arguments, result);
        break;

      case "deleteCollectData":
        YcProductPluginCollectData.deleteCollectData(arguments, result);
        break;

        // Ota升级 context
      case "deviceUpgrade":
        YcProductPluginOTA.deviceUpgrade(methodChannel, handler, activityPluginBinding.getActivity(), arguments, result);
        break;

        // MARK: - 表盘
      case "queryWatchFaceInfo":
        YcProductPluginWatchFace.queryWatchFaceInfo(arguments, result);
        break;

      case "changeWatchFace":
        YcProductPluginWatchFace.changeWatchFace(arguments, result);
        break;

      case "deleteWatchFace":
        YcProductPluginWatchFace.deleteWatchFace(arguments, result);
        break;

      case "installWatchFace":
        YcProductPluginWatchFace.installWatchFace(methodChannel, handler, arguments, result);
        break;

      case "queryDeviceCustomWatchFaceInfo":
        YcProductPluginWatchFace.queryDeviceCustomWatchFaceInfo(arguments, result);
        break;

      case "installCustomWatchFace":
        YcProductPluginWatchFace.installCustomWatchFace(context, methodChannel, handler, arguments, result);
        break;

      case "changeJieLiWatchFace":
        YcProductPluginWatchFace.changeJieLiWatchFace(arguments, result);
        break;

      case "deleteJieLiWatchFace":
        YcProductPluginWatchFace.deleteJieLiWatchFace(arguments, result);
        break;

      case "installJieLiWatchFace":
        YcProductPluginWatchFace.installJieLiWatchFace(methodChannel, handler, arguments, result);
        break;

      case "queryDeviceDisplayParametersInfo":
        YcProductPluginWatchFace.queryDeviceDisplayParametersInfo(arguments, result);
        break;

      case "installJieLiCustomWatchFace":
        YcProductPluginWatchFace.installJieLiCustomWatchFace(context, methodChannel, handler, arguments, result);
        break;

      case "updateJieLiDeviceContacts":
        YcProductPluginContact.updateJieLiDeviceContacts(context, arguments, result);
        break;

      case "queryJieLiDeviceContacts":
        YcProductPluginContact.queryJieLiDeviceContacts(context, arguments, result);
        break;

      case "updateDeviceContacts":
        YcProductPluginContact.updateDeviceContacts(context, arguments, result);
        break;
      
      case "getLogFilePath":
        YcProductPluginLogger.getLogFilePath(context, arguments, result);
        break;

    case "sendDeviceQuiteSleep":
        YcProductPluginAppControl.sendDeviceQuiteSleep(arguments, result);
        break;

      case "shareLogFile":
        YcProductPluginLogger.shareLogFile(context,activityPluginBinding, arguments, result);
        break;

      case "getJLDeviceLogFilePath":
        YcProductPluginLogger.getJLDeviceLogFilePath(context, arguments, result);
        break;
              
      case "getDeviceLogFilePath":
        YcProductPluginLogger.getDeviceLogFilePath(context, arguments, result);
        break;

      case "clearSDKLog":
        YcProductPluginLogger.clearSDKLog(context, arguments, result);
        break;

        //闹钟
      case "settingGetAllAlarm":
        YcProductPluginSetting.settingGetAllAlarm(arguments, result);
        break;
      case "settingAddAlarm":
        YcProductPluginSetting.settingAddAlarm(arguments, result);
        break;
      case "settingModfiyAlarm":
        YcProductPluginSetting.settingModfiyAlarm(arguments, result);
        break;
      case "updateCallAlerts":
        YcProductPluginSetting.updateCallAlerts(arguments, result);
        break;
      case "shutdown":
        YcProductPluginDevice.shutdown(result);
        break;
      case "startListening":
        YcProductPluginDevice.startListening(activityPluginBinding,context,result);
        break;
      case "appQuerySampleRate":
        YcProductPluginSetting.appQuerySampleRate(arguments,result);
        break;
      case "appConfigureSampleRate":
        YcProductPluginSetting.appConfigureSampleRate(arguments,result);
        break;
      case "appConfigureRealTimePpg":
        YcProductPluginSetting.appConfigureRealTimePpg(arguments,result);
        break;
      case "appQueryMems":
        YcProductPluginSetting.appQueryMems(arguments,result);
        break;
      case "appMemsSwitch":
        YcProductPluginSetting.appMemsSwitch(arguments,result);
        break;
      case "settingVibrationIntensity":
        YcProductPluginSetting.settingVibrationIntensity(arguments,result);
        break;
      default:
        result.notImplemented();

    }
  }


  /**
   * 设置监听
   */
  private void setupObserver(Handler handler, EventChannel.EventSink eventSink) {

    // 蓝牙状态
    YcProductPluginDevice.setupDeviceStateObserver(handler, eventSink);

    // 设备操作
    YcProductPluginDeviceControl.setupDeviceControlObserver(handler, eventSink);

    // 实时数据
    YcProductPluginRealData.setupDeviceRealDataObserver(handler, eventSink);
  }


  /**
   * 转换Flutter插件的状态码
   *
   * @param code
   * @return
   */
  public static int convertPluginState(int code) {

    if (0xFC == (code & 0xFF) || 0xFD == (code & 0xFF)) {

      return YcProductPluginFlutterType.PluginState.unavailable;

    } else if (0 == code) {

      return YcProductPluginFlutterType.PluginState.succeed;

    } else {

      return YcProductPluginFlutterType.PluginState.failed;

    }
  }

}


